Added changelog. (#1555)
OpenZeppelin/openzeppelin-solidity: CHANGELOG.md

> ### 2.1.0 (unreleased)

Reviewer: In the changelogs up until now I've included with each item the relevant PR number. I think it's a nice detail for people who want to look into the details behind a decision. In this text file it would be more cumbersome to do because `#1234` is not automatically converted into a link. The information will be there in the git blame, but that's quite a bit harder to find.

What do you think?

Author: I'd go with Mardown links, since those will be properly formatted in both GitHub and Discourse. I didn't originally add them due to laziness, but they are a worthy addition.

Reviewer: I love the result but I'm concerned it'll be cumbersome for contributors. Let's see how it works.

> New features:

Reviewer: I think a header would look nicer here. I think it'll also enable linking to a particular release (`/CHANGELOG.md#2.1.0`).

I'd consider the following formatting:

```
# Changelog

## 2.1.0 (unreleased)

### New features

- `WhitelistCrowdsale`
```

<details>
<summary>Rendered</summary>

# Changelog

## 2.1.0 (unreleased)

### New features

- `WhitelistCrowdsale`
</details>

Author: Good thinking!

>  * `WhitelistCrowdsale`: a crowdsale where only whitelisted accounts (`WhitelistedRole`) can purchase tokens. This attribute is given by a series of whitelisters (`WhitelisterRole`), who can also remove them. Similar to the pre-2.0 `WhitelistedCrowdsale`.

Reviewer: Wording.

```suggestion
 * `WhitelistCrowdsale`: a crowdsale where only whitelisted accounts (`WhitelistedRole`) can purchase tokens. Adding or removing accounts from the whitelist is done by whitelisters (`WhitelisterRole`). Similar to the pre-2.0 `WhitelistedCrowdsale`.
```

>  * `RefundablePostDeliveryCrowdsale`: replacement for the (deprecated) `RefundableCrowdsale`: tokens are only granted once the crowdsale ends (if it meets its goal).

Reviewer: Too many colons.

```suggestion
 * `RefundablePostDeliveryCrowdsale`: replacement for `RefundableCrowdsale` (deprecated, see below) where tokens are only granted once the crowdsale ends (if it meets its goal).
```

>  * `ERC20`: `transferFrom` and `_burnFrom ` now emit `Approval` events, improving dApp support.

Reviewer: I found many of these descriptions to be too vague. What does "improving dApp support" mean?

I'd prefer something like this. What do you think?
```suggestion
 * `ERC20`: `transferFrom` and `_burnFrom ` now emit `Approval` events, to represent the token's state comprehensively through events.
```

Author: I do think my original description was too vague, but I'd like for us to explain _why_ 'representing the token's state comprehensively through events' is important.

Reviewer: I agree. Do you think the changelog is the right place for that? Perhaps we should have a link to "read more"?

>  * `ERC721`: added `_burn(uint256 tokenId)`, replacing the (deprecated) old function.

Reviewer: ```suggestion
 * `ERC721`: added `_burn(uint256 tokenId)`, replacing the similar deprecated function (see below).
```

>  * `ERC721`: added `_tokensOfOwner(address owner)`, allowing for more extensive `ERC721` customization.

Reviewer: More precise description.

```suggestion
 * `ERC721`: added `_tokensOfOwner(address owner)`, allowing to internally retrieve the array of an account's owned tokens.
```

Improvements:

>  * All contracts now require solc v0.4.24 or higher.

Reviewer: It was only one contract, right? I'd be more explicit about that (i.e. what changed exactly) because otherwise users will wonder if this affects them in any way.

Author: Yup. I actually added this thinking we'd go up to 0.4.25, but we haven't discussed that :stuck_out_tongue:

Reviewer: Heh, yeah. IMO let's wait until next release when we figure out a strategy for dealing with solc updates.

 * Now conforming to a 4-space indentation code style.
 * `ERC20`: more gas efficient due to removed redundant `require`s.
 * `ERC721`: fixed a bug that prevented internal data structures from being properly cleaned, missing potential gas refunds.
 * `ERC721`: general gas savings on `transferFrom`, `_mint` and `_burn`, due to redudant `require`s and `SSTORE`s.

Bugfixes:

> Breaking changes:

Reviewer: None of these are breaking changes, and the only breaking changes we introduced are not listed: removing the internal `_addTokenTo` and `_removeTokenFrom` functions in ERC721. I know those where not meant to be part of the API, but I'd still add it here.

As for these two items, I'd list them under a "Deprecations" header.

Author: I added them here to avoid having yet another category, but that may be less confusing for users.
